crates/utils/src/interval_map/tests.rs

```diff
@@ -0,0 +1,304 @@
+    rng: StdRng,
+    unique: HashSet<Interval<i32>>,
+    limit: i32,
+}
+
+impl IntervalGenerator {
+    fn new(seed: [u8; 32]) -> Self {
+        const LIMIT: i32 = 1000;
+        Self {
+            rng: SeedableRng::from_seed(seed),
+            unique: HashSet::new(),
+            limit: LIMIT,
+        }
+    }
+
+    fn next(&mut self) -> Interval<i32> {
+        let low = self.rng.gen_range(0..self.limit - 1);
+        let high = self.rng.gen_range((low + 1)..self.limit);
+        Interval::new(low, high)
+    }
+
+    fn next_unique(&mut self) -> Interval<i32> {
+        let mut interval = self.next();
+        while self.unique.contains(&interval) {
+            interval = self.next();
+        }
+        self.unique.insert(interval.clone());
+        interval
+    }
+
+    fn next_with_range(&mut self, range: i32) -> Interval<i32> {
+        let low = self.rng.gen_range(0..self.limit - 1);
+        let high = self
+            .rng
+            .gen_range((low + 1)..self.limit.min(low + 1 + range));
+        Interval::new(low, high)
+    }
+}
+
+impl<V> IntervalMap<i32, V> {
+    fn check_max(&self) {
+        let _ignore = Self::check_max_inner(&self.root);
+    }
+
+    fn check_max_inner(x: &NodeRef<i32, V>) -> i32 {
+        if x.is_sentinel() {
+            return 0;
+        }
+        let l_max = Self::check_max_inner(&x.left_owned());
+        let r_max = Self::check_max_inner(&x.right_owned());
+        let max = x.interval(|i| i.high.max(l_max).max(r_max));
+        assert_eq!(x.max_owned(), max);
+        max
+    }
+
+    /// 1. Every node is either red or black.
+    /// 2. The root is black.
+    /// 3. Every leaf (NIL) is black.
+    /// 4. If a node is red, then both its children are black.
+    /// 5. For each node, all simple paths from the node to descendant leaves contain the
+    /// same number of black nodes.
+    fn check_rb_properties(&self) {
+        assert!(matches!(self.root.color(), Color::Black));
+        Self::check_children_color(&self.root);
+        Self::check_black_height(&self.root);
+    }
+
+    fn check_children_color(x: &NodeRef<i32, V>) {
+        if x.is_sentinel() {
+            return;
+        }
+        Self::check_children_color(&x.left_owned());
+        Self::check_children_color(&x.right_owned());
+        if x.is_red() {
+            assert!(x.left(|l| matches!(l.color(), Color::Black)));
+            assert!(x.right(|r| matches!(r.color(), Color::Black)));
+        }
+    }
+
+    fn check_black_height(x: &NodeRef<i32, V>) -> usize {
+        if x.is_sentinel() {
+            return 0;
+        }
+        let lefth = Self::check_black_height(&x.left_owned());
+        let righth = Self::check_black_height(&x.right_owned());
+        assert_eq!(lefth, righth);
+        if x.is_black() {
+            return lefth + 1;
+        }
+        lefth
+    }
+}
+
+fn with_map_and_generator<V>(test_fn: impl FnOnce(IntervalMap<i32, V>, IntervalGenerator) + Clone) {
```

Comment:
Can we use `madsim` to implement this method?

Reply:
The `SeedableRng` is already deterministic, so there's no need to use madsim.